File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.js
var lengthOfLongestSubstring = function(s) {
    const map = new Map()
    let left = 0, right = 0;
    let maxLen = 0
    
    for (right; right < s.length; right++) {
        const cur = s[right]
        
        if (map.has(cur)) left = Math.max(left, map.get(cur) + 1)
        map.set(cur, right)
        maxLen = Math.max(maxLen, right - left + 1)
    }
    
    return maxLen
};  

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp

class Solution {

public:
    int lengthOfLongestSubstring(std::string str) {
        
        if (str.length() == 0)
            return 0;

        int left = 0;  int right = 0;

        std::map<char, int> m;
        int maximum = 1;
        int len = 0;

        for(left ; left < str.length() ; left++)
        {
            // std::cout << "left : " << left << "\n";
            for (right = left; right < str.length(); right++)
            {

                if ( m.count( str[right]) == true )
                {
                    maximum = std::max(len, maximum);
                    len = 0;
                    m.clear();
                    break;
                }
                else
                {
                    len++;
                    m.insert(std::pair<char, int>(str[right], right));
                }
            }
        }
        return maximum;
    }
};
